File: DraftRightWindows/DraftRightWindows/Services/GoogleOAuth.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Net;
using System.Net.Http;
using System.Net.Sockets;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace DraftRightWindows.Services;

/// <summary>
/// Google OAuth 2.0 for an installed Windows app — loopback redirect + PKCE.
///
/// Per Google's "Installed app" guidance, this is a *public* client: the
/// client_secret ships with the binary and is not security-critical; PKCE is
/// the actual proof-of-possession. Custom URL schemes (used by the macOS/iOS
/// clients) aren't needed on Windows — Google supports listening on
/// <c>http://127.0.0.1:&lt;ephemeral-port&gt;</c> instead, which avoids registering
/// a protocol handler in the registry.
///
/// Flow:
///   1. Generate PKCE verifier + S256 challenge.
///   2. Open an HttpListener on a free loopback port.
///   3. Launch the system browser at the Google consent screen.
///   4. Google redirects back to <c>http://127.0.0.1:PORT/?code=...</c>; we
///      capture the code, respond with a friendly HTML page so the tab
///      doesn't look broken.
///   5. Exchange the code (with the verifier) for an <c>id_token</c>.
///
/// The returned id_token is passed to the backend's <c>/auth/social</c>
/// endpoint, which verifies it via Google's tokeninfo and creates/links
/// the user.
/// </summary>
public static class GoogleOAuth
{
    // Desktop-type OAuth client (project 22951518033). Public client: the
    // "secret" is distributed with every Windows copy of the app — not a real
    // secret. PKCE is the actual proof-of-possession. Kept out of git for
    // hygiene (see GoogleOAuthConfig.cs.template).
    private const string ClientId     = "22951518033-oq7okrvvbb26eqsb7c0avsb1ic165ole.apps.googleusercontent.com";
    private static string ClientSecret => GoogleOAuthConfig.ClientSecret;
    private const string AuthEndpoint  = "https://accounts.google.com/o/oauth2/v2/auth";
    private const string TokenEndpoint = "https://oauth2.googleapis.com/token";
    private const string Scopes        = "openid email profile";

    /// <summary>Run the full flow → return Google id_token. Throws on cancel or error.</summary>
    public static async Task<string> AuthenticateAsync()
    {
        var verifier  = PkceVerifier();
        var challenge = PkceChallenge(verifier);
        var state     = RandomToken(16);

        var port        = FindFreeLoopbackPort();
        var redirectUri = $"http://127.0.0.1:{port}/";

        using var listener = new HttpListener();
        listener.Prefixes.Add(redirectUri);
        listener.Start();

        var authUrl = AuthEndpoint
            + "?response_type=code"
            + "&client_id="              + Uri.EscapeDataString(ClientId)
            + "&redirect_uri="           + Uri.EscapeDataString(redirectUri)
            + "&scope="                  + Uri.EscapeDataString(Scopes)
            + "&code_challenge="         + challenge
            + "&code_challenge_method=S256"
            + "&state="                  + state;

        Process.Start(new ProcessStartInfo { FileName = authUrl, UseShellExecute = true });

        // Wait for Google to redirect back. listener.GetContextAsync blocks
        // until the browser hits the loopback — no timeout here because the
        // user may take their time on the consent screen; the caller decides
        // whether to surface a cancel UI.
        var ctx = await listener.GetContextAsync();
        var query = ParseQuery(ctx.Request.Url!.Query);

        // Always show the user a clean "you can close this" page, even on
        // error — a blank browser tab looks worse than the failure itself.
        WriteHtmlResponse(ctx,
            "<html><body style='font:14px sans-serif;text-align:center;padding:40px'>"
            + "<h2>Sign-in complete</h2>"
            + "<p>You can close this window and return to DraftRight.</p>"
            + "</body></html>");

        if (query.TryGetValue("error", out var err) && !string.IsNullOrEmpty(err))
            throw new InvalidOperationException("Google sign-in error: " + err);
        if (!query.TryGetValue("code", out var code) || string.IsNullOrEmpty(code))
            throw new InvalidOperationException("Google sign-in returned no code");
        if (!query.TryGetValue("state", out var gotState) || gotState != state)
            throw new InvalidOperationException("OAuth state mismatch (possible CSRF)");

        return await ExchangeCodeForIdTokenAsync(code, verifier, redirectUri);
    }

    private static async Task<string> ExchangeCodeForIdTokenAsync(string code, string verifier, string redirectUri)
    {
        using var http = new HttpClient();
        var form = new FormUrlEncodedContent(new Dictionary<string, string>
        {
            ["code"]          = code,
            ["client_id"]     = ClientId,
            ["client_secret"] = ClientSecret,
            ["redirect_uri"]  = redirectUri,
            ["grant_type"]    = "authorization_code",
            ["code_verifier"] = verifier,
        });

        var resp = await http.PostAsync(TokenEndpoint, form);
        var json = await resp.Content.ReadAsStringAsync();
        if (!resp.IsSuccessStatusCode)
            throw new InvalidOperationException($"Google token exchange failed: {(int)resp.StatusCode} {json}");

        using var doc = JsonDocument.Parse(json);
        if (!doc.RootElement.TryGetProperty("id_token", out var idTok))
            throw new InvalidOperationException("Token response had no id_token");
        var s = idTok.GetString();
        if (string.IsNullOrEmpty(s)) throw new InvalidOperationException("Empty id_token in token response");
        return s;
    }

    // --- helpers ---

    /// <summary>Bind a TcpListener to port 0 → OS picks a free port → release.</summary>
    private static int FindFreeLoopbackPort()
    {
        var l = new TcpListener(IPAddress.Loopback, 0);
        l.Start();
        try { return ((IPEndPoint)l.LocalEndpoint).Port; }
        finally { l.Stop(); }
    }

    private static string PkceVerifier()
    {
        var bytes = new byte[32];
        RandomNumberGenerator.Fill(bytes);
        return Base64Url(bytes);
    }

    private static string PkceChallenge(string verifier)
    {
        var hash = SHA256.HashData(Encoding.UTF8.GetBytes(verifier));
        return Base64Url(hash);
    }

    private static string RandomToken(int byteLen)
    {
        var bytes = new byte[byteLen];
        RandomNumberGenerator.Fill(bytes);
        return Base64Url(bytes);
    }

    /// <summary>base64url without padding (the encoding OAuth/PKCE require).</summary>
    private static string Base64Url(byte[] data) =>
        Convert.ToBase64String(data).TrimEnd('=').Replace('+', '-').Replace('/', '_');

    private static Dictionary<string, string> ParseQuery(string query)
    {
        var dict = new Dictionary<string, string>(StringComparer.Ordinal);
        if (string.IsNullOrEmpty(query)) return dict;
        if (query[0] == '?') query = query.Substring(1);
        foreach (var pair in query.Split('&'))
        {
            if (string.IsNullOrEmpty(pair)) continue;
            var eq = pair.IndexOf('=');
            if (eq < 0) { dict[Uri.UnescapeDataString(pair)] = ""; continue; }
            var key = Uri.UnescapeDataString(pair.Substring(0, eq));
            var val = Uri.UnescapeDataString(pair.Substring(eq + 1));
            dict[key] = val;
        }
        return dict;
    }

    private static void WriteHtmlResponse(HttpListenerContext ctx, string html)
    {
        var bytes = Encoding.UTF8.GetBytes(html);
        ctx.Response.ContentType = "text/html; charset=utf-8";
        ctx.Response.ContentLength64 = bytes.Length;
        ctx.Response.OutputStream.Write(bytes, 0, bytes.Length);
        ctx.Response.OutputStream.Close();
    }
}

File: DraftRightLinux/draftright/services/google_oauth.py
"""Google sign-in for a native Linux app (#97).

Implements the OAuth 2.0 native-app flow from RFC 8252: PKCE plus a **loopback
redirect**.  macOS uses an iOS-type client with a reversed-scheme redirect
(``com.googleusercontent.apps.<id>:/oauth2callback``); that scheme is only
accepted by iOS-type clients, so Linux uses the loopback form instead, which is
what Google's "Desktop app" client type accepts.

    1. Generate a PKCE verifier + S256 challenge and a random ``state``.
    2. Bind 127.0.0.1 on an ephemeral port; that URL is the redirect_uri.
    3. Open the system browser at Google's authorize endpoint.
    4. Google redirects back to the loopback server with ``code``.
    5. Exchange code + verifier at the token endpoint → ``id_token``.

The ``id_token`` then goes to ``POST /auth/social``, which verifies it with
Google server-side.  No client secret is involved: a native app is a *public*
client and PKCE is the proof-of-possession.
"""

from __future__ import annotations

import base64
import hashlib
import http.server
import logging
import secrets
import threading
import urllib.parse
import webbrowser

import requests

from draftright import config

log = logging.getLogger(__name__)

_SUCCESS_HTML = b"""<!doctype html><html><head><meta charset="utf-8">
<title>DraftRight</title></head><body style="font-family:system-ui;text-align:center;padding:3rem">
<h2>You're signed in</h2><p>You can close this tab and return to DraftRight.</p>
</body></html>"""

_FAILURE_HTML = b"""<!doctype html><html><head><meta charset="utf-8">
<title>DraftRight</title></head><body style="font-family:system-ui;text-align:center;padding:3rem">
<h2>Sign-in failed</h2><p>Return to DraftRight and try again.</p>
</body></html>"""


class GoogleOAuthError(RuntimeError):
    """Sign-in did not complete."""


class _CallbackHandler(http.server.BaseHTTPRequestHandler):
    """Captures the single OAuth redirect, then lets the server stop."""

    # Set by the server instance.
    result: dict = {}

    def do_GET(self):  # noqa: N802 — BaseHTTPRequestHandler API
        parsed = urllib.parse.urlparse(self.path)
        params = urllib.parse.parse_qs(parsed.query)
        # parse_qs gives lists; collapse to the first value of each.
        self.server.result = {k: v[0] for k, v in params.items() if v}

        ok = "code" in self.server.result
        self.send_response(200)
        self.send_header("Content-Type", "text/html; charset=utf-8")
        self.end_headers()
        self.wfile.write(_SUCCESS_HTML if ok else _FAILURE_HTML)

    def log_message(self, *_args):
        """Silence BaseHTTPRequestHandler's stderr logging."""


class _LoopbackServer(http.server.HTTPServer):
    result: dict = {}


def _pkce_pair() -> tuple[str, str]:
    """Return (verifier, S256 challenge), base64url without padding."""
    verifier = base64.urlsafe_b64encode(secrets.token_bytes(64)).rstrip(b"=").decode()
    digest = hashlib.sha256(verifier.encode("ascii")).digest()
    challenge = base64.urlsafe_b64encode(digest).rstrip(b"=").decode()
    return verifier, challenge


def sign_in(timeout: float = config.GOOGLE_OAUTH_TIMEOUT) -> str:
    """Run the browser sign-in flow and return a Google ``id_token``.

    Blocks until the user finishes in the browser, so call it off the GTK main
    thread.

    Raises:
        GoogleOAuthError: cancelled, timed out, or the exchange failed.
    """
    verifier, challenge = _pkce_pair()
    state = secrets.token_urlsafe(24)

    # Port 0 → the kernel picks a free port; Google allows any port on the
    # loopback interface for native clients (RFC 8252 §7.3).
    server = _LoopbackServer(("127.0.0.1", 0), _CallbackHandler)
    server.result = {}
    redirect_uri = f"http://127.0.0.1:{server.server_address[1]}"

    thread = threading.Thread(target=server.handle_request, daemon=True)
    thread.start()

    query = urllib.parse.urlencode(
        {
            "client_id": config.google_client_id(),
            "response_type": "code",
            "redirect_uri": redirect_uri,
            "scope": config.GOOGLE_OAUTH_SCOPES,
            "code_challenge": challenge,
            "code_challenge_method": "S256",
            "state": state,
            # Always show the picker; otherwise a signed-in browser silently
            # reuses an account the user may not have meant to link.
            "prompt": "select_account",
        }
    )
    auth_url = f"{config.GOOGLE_AUTH_ENDPOINT}?{query}"
    log.info("Opening the browser for Google sign-in…")
    if not webbrowser.open(auth_url):
        server.server_close()
        raise GoogleOAuthError("Couldn't open a browser for Google sign-in.")

    thread.join(timeout)
    result = dict(server.result)
    server.server_close()

    if not result:
        raise GoogleOAuthError("Timed out waiting for Google sign-in.")
    if "error" in result:
        raise GoogleOAuthError(f"Google returned: {result['error']}")
    if result.get("state") != state:
        # Mismatched state means the response isn't ours — refuse it.
        raise GoogleOAuthError("Sign-in state mismatch; please try again.")
    code = result.get("code")
    if not code:
        raise GoogleOAuthError("Google did not return an authorization code.")

    return _exchange_code(code, verifier, redirect_uri)


def _exchange_code(code: str, verifier: str, redirect_uri: str) -> str:
    """Swap the authorization code for an id_token."""
    try:
        response = requests.post(
            config.GOOGLE_TOKEN_ENDPOINT,
            data={
                "client_id": config.google_client_id(),
                # Required by Google for Desktop-type clients; see config.
                "client_secret": config.google_client_secret(),
                "code": code,
                "code_verifier": verifier,
                "grant_type": "authorization_code",
                "redirect_uri": redirect_uri,
            },
            timeout=config.API_TIMEOUT,
        )
    except requests.RequestException as exc:
        raise GoogleOAuthError(f"Couldn't reach Google: {exc}") from exc

    if response.status_code >= 400:
        raise GoogleOAuthError(
            f"Google rejected the sign-in ({response.status_code}): "
            f"{response.text[:200]}"
        )

    id_token = response.json().get("id_token")
    if not id_token:
        raise GoogleOAuthError("Google's response contained no id_token.")
    return id_token
